builtin/logical/database/rotation.go

```diff
@@ -0,0 +1,522 @@
+			RoleName:          input.RoleName,
+			Username:          config.Username,
+			NewPassword:       config.Password,
+			OldPassword:       input.Role.StaticAccount.Password,
+			LastVaultRotation: input.Role.StaticAccount.LastVaultRotation,
+		})
+		if err != nil {
+			return output, errwrap.Wrapf("error writing WAL entry: {{err}}", err)
+		}
+	}
+
+	_, password, err := db.SetCredentials(ctx, input.Role.Statements, config)
+	if err != nil {
+		b.CloseIfShutdown(db, err)
+		return output, errwrap.Wrapf("error setting credentials: {{err}}", err)
+	}
+
+	if newPassword != password {
+		return output, errors.New("mismatch passwords returned")
+	}
+
+	// Store updated role information
+	lvr = time.Now()
+	input.Role.StaticAccount.LastVaultRotation = lvr
+	input.Role.StaticAccount.Password = password
+	output.RotationTime = lvr
+
+	entry, err := logical.StorageEntryJSON(databaseStaticRolePath+input.RoleName, input.Role)
+	if err != nil {
+		return output, err
+	}
+	if err := s.Put(ctx, entry); err != nil {
+		return output, err
+	}
+
+	// Cleanup WAL after successfully rotating and pushing new item on to queue
+	if err := framework.DeleteWAL(ctx, s, output.WALID); err != nil {
+		merr = multierror.Append(merr, err)
+		return output, merr
+	}
+
+	// The WAL has been deleted, return new setStaticAccountOutput without it
+	return &setStaticAccountOutput{RotationTime: lvr}, merr
+}
+
+// initQueue preforms the necessary checks and initializations needed to preform
+// automatic credential rotation for roles associated with static accounts. This
+// method verifies if a queue is needed (primary server or local mount), and if
+// so initializes the queue and launches a go-routine to periodically invoke a
+// method to preform the rotations.
+//
+// initQueue is invoked by the Factory method in a go-routine. The Factory does
+// not wait for success or failure of it's tasks before continuing. This is to
+// avoid blocking the mount process while loading and evaluating existing roles,
+// etc.
+func (b *databaseBackend) initQueue(ctx context.Context, conf *logical.BackendConfig) {
+	// Verify this mount is on the primary server, or is a local mount. If not, do
+	// not create a queue or launch a ticker. Both processing the WAL list and
+	// populating the queue are done sequentially and before launching a
+	// go-routine to run the periodic ticker.
+	replicationState := conf.System.ReplicationState()
+	if (conf.System.LocalMount() || !replicationState.HasState(consts.ReplicationPerformanceSecondary)) &&
+		!replicationState.HasState(consts.ReplicationDRSecondary) &&
+		!replicationState.HasState(consts.ReplicationPerformanceStandby) {
+		b.Logger().Info("initializing database rotation queue")
+
+		// Poll for a PutWAL call that does not return a "read-only storage" error.
+		// This ensures the startup phases of loading WAL entries from any possible
+		// failed rotations can complete without error when deleting from storage.
+	READONLY_LOOP:
+		for {
```

Comment:
I think there needs to be a few safeguards in this loop.  First, I think you need to check the context to make sure it hasn't been cancelled to break out of this loop.  Also, since this will block startup of the plugin, I think it may make sense to put in some sane max timeout.

Reply:
> I think you need to check the context to make sure it hasn't been cancelled to break out of this loop

Great point, thanks. I added a `switch` statement to listen for `Done` in ae5563c38.

> since this will block startup of the plugin

`initQueue` is ran in a background go-routine, kicked off in `Factory`, so it should not block the startup of the plugin